tests/tests_combination_absent.yml

```diff
@@ -0,0 +1,155 @@
+    __test_inputfiles_conf: /etc/rsyslog.d/90-input-files-{{ __test_tag }}.conf
+    __test_inputfiles_dir: /var/log/inputdirectory
+    __test_basics_conf: 11-input-basics-basic_input.conf
+
+  tasks:
+    - name: Generate an input log dir {{ __test_inputfiles_dir }}
+      file:
+        path: "{{ __test_inputfiles_dir }}"
+        state: directory
+        mode: 0700
+
+    - name: deploy config to output into local files
+      vars:
+        logging_outputs:
+          - name: files_test0
+            type: files
+            severity: info
+            exclude:
+              - authpriv.none
+              - auth.none
+              - cron.none
+              - mail.none
+            path: /var/log/messages
+          - name: files_test1
+            type: files
+            facility: authpriv,auth
+            path: /var/log/secure
+          - name: forwards_severity_and_facility
+            type: forwards
+            facility: local1
+            severity: info
+            protocol: tcp
+            target: host.domain
+            port: 1514
+          - name: forwards_facility_only
+            type: forwards
+            facility: local2
+            protocol: tcp
+            target: host.domain
+            port: 2514
+        logging_inputs:
+          - name: basic_input
+            type: basics
+            rsyslog_imjournal_ratelimit_burst: 33333
+          - name: "{{ __test_tag }}"
+            type: files
+            rsyslog_input_log_path: "{{ __test_inputfiles_dir }}/*.log"
+        logging_flows:
+          - name: flow_0
+            inputs: [basic_input]
+            outputs: [files_test0, files_test1, forwards_severity_and_facility, forwards_facility_only]
+          - name: flow_1
+            inputs: ["{{ __test_tag }}"]
+            outputs: [files_test0, files_test1]
+      include_role:
+        name: linux-system-roles.logging
+
+    - include_tasks: set_rsyslog_variables.yml
+
+    # notify restart rsyslogd is executed at the end of this test task.
+    # thus we have to force to invoke handlers
+    - name: Force all notified handlers to run at this point, not waiting for normal sync points
+      meta: flush_handlers
+
+    - name: Check rsyslog version
+      assert:
+        that: __rsyslog_version != "N/A" and __rsyslog_version != "0.0.0"
+
+    - name: Check rsyslog.conf size
+      assert:
+        that: rsyslog_conf_stat.stat.size < 1000
+
+    - name: Check file counts in rsyslog.d
+      assert:
+        that: rsyslog_d_file_count.matched >= 11
+
+    - name: generate test config file
+      copy:
+        dest: "/tmp/{{ __test_basics_conf }}"
+        content: |-
+          # GENERATED BASICS CONFIG FILE
+          if ($inputname == "imjournal") then {
+              call files_test0
+          }
+      delegate_to: localhost
+
+    - name: Update it to the target
+      copy:
+        src: "/tmp/{{ __test_basics_conf }}"
+        dest: "/etc/rsyslog.d/{{ __test_basics_conf }}"
```

Comment:
Do you need to create a local file first, before you copy it?  Can you remove the `name: generate test config file` task, and just do
```yaml

    - name: generate test config file
      copy:
        dest: "/etc/rsyslog.d/{{ __test_basics_conf }}"
        content: |-
          # GENERATED BASICS CONFIG FILE
          if ($inputname == "imjournal") then {
              call files_test0
          }
```
?

Reply:
Thank you, @richm. Fixing it.
Strangely, CI tests are failing on RHEL/CentOS-7 (not on 8 or Fedoras...) Invetigating it.

Reply:
@richm, this might be really problematic.
```
Questionable case) XX-filename.conf exists; instead of having the
item with the absent state, delete the item from logging_inputs or
outputs. Then, the existing XX-filename.conf remains untouched.
```
In tests_combination_absent.yml, it does a full deployment in the first run; in the second run's config, some dict items are set to `absent` and some dict items are removed. The second run successfully removes the config files to be generated from the `absent` items, but it leaves config files from the dict items removed in the second run. It causes this error.
`May 28 22:30:44 ibm-p8-kvm-03-guest-02.virt.pnr.lab.eng.rdu2.redhat.com rsyslogd[14038]: error during config processing: ruleset 'files_test1' cannot be found  [v8.24.0-52.el7 try http://www.rsyslog.com/e/2207 ]`

Actually, there is another issue in RHEL/CentOS-7. systemd package puts listen.conf in /etc/rsyslog.d.
```
# rpm -qf /etc/rsyslog.d/listen.conf 
systemd-219-62.el7_6.9.x86_64
```
It expects imuxsock, but we set it off.
```
# cat /etc/rsyslog.d/listen.conf 
$SystemLogSocketName /run/systemd/journal/syslog
```
It makes rsyslog issue this error (although rsyslog is up and running.)
`May 28 22:30:44 ibm-p8-kvm-03-guest-02.virt.pnr.lab.eng.rdu2.redhat.com rsyslogd[14038]: command 'SystemLogSocketName' is currently not permitted - did you already set it via a RainerScript command (v6+ config)? [v8.24.0-52.el7 try http://www.rsyslog.com/e/2222 ]`

The easiest way to solve the above 2 issues is to set `logging_purge_confs: true` in the second run, which cleans up all existing files in /etc/rsyslog.d and starts from scratch. But it may cause some other issue such as there's another config file `21-cloudinit.conf` placed by `cloud-init-18.2-1.el7.centos.2.x86_64` which we may not want to delete...

For instance, is it acceptable to have a black list for the task handling logging_purge_confs? Can we put listen.conf and config files starting with ##-input- or ##-output- which are ours? Is it too ugly?
```
# cat 21-cloudinit.conf 
# Log cloudinit generated log messages to file
:syslogtag, isequal, "[CLOUDINIT]" /var/log/cloud-init.log
```

Reply:
Please note that now all the CI tests are failing since I added the systemctl status error check to tests_combination_absent.yml.

Reply:
> The second run successfully removes the config files to be generated from the absent items, but it leaves config files from the dict items removed in the second run.

I'm not sure what you mean.

> listen.conf
```
# cat /etc/rsyslog.d/listen.conf 
$SystemLogSocketName /run/systemd/journal/syslog
```
so systemd adds a file that assumes rsyslog is loading the `imuxsock` module?

one way to solve it would be to always load the `imuxsock` module.  I guess that's part of the rsyslog "public API" that we can't easily change, especially for EL7.  Do we have the same problem on EL8?

Reply:
> Please note that now all the CI tests are failing since I added the systemctl status error check to tests_combination_absent.yml.

Then I guess the systemctl status error check is working  e.g. https://fedorapeople.org/groups/linuxsystemroles/logs/linux-system-roles-logging-pull-linux-system-roles_logging-124-3f46aca-fedora-32-20200529-000038/artifacts/ansible.log
```
TASK [Check rsyslog errors] ****************************************************
task path: /tmp/tmpgf02lvy_/tests/tests_combination_absent.yml:160
fatal: [/cache/Fedora-Cloud-Base-32-1.6.x86_64.qcow2]: FAILED! => {"changed": true, "cmd": ["systemctl", "status", "rsyslog"], ... systemd[1]: Starting System Logging Service...\nMay 29 00:10:22 ibm-p8-kvm-03-guest-02.virt.pnr.lab.eng.rdu2.redhat.com rsyslogd[16994]: error during parsing file /etc/rsyslog.conf, on or before line 6: ruleset 'files_test1' cannot be found 
```
that looks like a legitimate error message.